### PR TITLE
fix(optimizer-geometry): use exact power-of-two rescaling in flad_noise_component_transaction

### LINKED ISSUE #2393: flad_noise_component_transaction drops the whole gradient component when the squared norm underflows
## Summary

`flad_noise_component_transaction` in `alberta_framework/evaluation/optimizer_geometry.py` silently returns the perturbation unchanged, with `valid=True`, whenever the gradient's squared norm underflows to zero in float32. The returned "noise component" then still contains the entire component along the gradient — the one quantity the function exists to remove.

This is not a small numerical error. The component that should be removed is `O(1)` relative to the perturbation, and it is dropped in full.

## Reproduction

```python
import jax
import jax.numpy as jnp
import numpy as np

from alberta_framework.evaluation.optimizer_geometry import (
    flad_noise_component_transaction,
)

DELTA = np.array([1.0, -2.0, 0.5, 3.0], dtype=np.float32)
GRAD = np.array([2.0, 1.0, -1.0, 0.5], dtype=np.float32)


def reference(delta, gradient):
    d = np.asarray(delta, dtype=np.float64)
    g = np.asarray(gradient, dtype=np.float64)
    denom = float(g @ g)
    return d if denom == 0.0 else d - g * (float(g @ d) / denom)


for scale in (1.0, 1e-10, 1e-19, 1e-20, 1e-25, 1e-30):
    gradient = jnp.asarray(GRAD * np.float32(scale))
    safe, valid = jax.jit(flad_noise_component_transaction)(jnp.asarray(DELTA), gradient)
    err = np.max(np.abs(np.asarray(safe, dtype=np.float64) - reference(DELTA, GRAD * np.float32(scale))))
    print(f"scale={scale:8.1e} g@g={float(jnp.vdot(gradient, gradient).real):12.5e} "
          f"valid={bool(valid)} max_err={err:12.5e}")
```

Measured on JAX 0.11.0, CPU backend, x64 disabled, identical eagerly and under `jax.jit`:

| gradient scale | `g @ g` (float32) | `g @ delta` (float32) | `valid` | max error vs float64 |
| --- | --- | --- | --- | --- |
| 1.0 | 6.25000e+00 | 1.00000e+00 | True | 8.58307e-08 |
| 1e-10 | 6.25000e-20 | 1.00000e-10 | True | 8.58307e-08 |
| 1e-19 | 6.25000e-38 | 1.00000e-19 | True | **6.09524e-02** |
| 1e-20 | **0.00000e+00** | 1.00000e-20 | True | **3.20000e-01** |
| 1e-25 | **0.00000e+00** | 1.00000e-25 | True | **3.20000e-01** |
| 1e-30 | **0.00000e+00** | 1.00000e-30 | True | **3.20000e-01** |

The component along the gradient is exactly `3.20000e-01` for this input, so at 1e-20 and below the function removes none of it. At 1e-19 the denominator is a barely-normal `6.25e-38` and roughly 19% of the projection is already lost.

## Why it happens

```python
squared_norm = jnp.vdot(direction, direction).real
numerator = jnp.vdot(direction, delta).real
active = squared_norm > 0.0
denominator = jnp.where(active, squared_norm, jnp.ones_like(squared_norm))
coefficient = numerator / denominator
projection = direction * coefficient * active.astype(delta.dtype)
```

`active` is the mechanism-off control: `GEOMETRY_PROTOCOL["mechanism_off"]` is `"empty_basis_or_zero_gradient_exact_reduction"`, and a genuinely zero gradient must reduce to returning `delta` exactly. But `squared_norm > 0.0` cannot distinguish a zero gradient from a gradient whose *squares* underflowed. For float32 entries at or below about 1e-20 the squares land below the subnormal floor and the accumulation is exactly zero, so a perfectly well-conditioned gradient direction takes the zero-gradient branch.

The information is not lost. `numerator` survives at every scale in the table above — `1e-20` and `1e-30` are both exactly representable. Only the denominator dies. Unlike the flushed-operand case in #2391, this one is fully recoverable.

## Suggested fix

The projection onto `span(gradient)` is invariant to rescaling `gradient`, so rescaling by an exact power of two is an exact reformulation rather than an approximation — the same shape as the divisor repair in #2380:

```python
_, exponent = jnp.frexp(jnp.max(jnp.abs(direction)))
scaled = jnp.ldexp(direction, -exponent)
squared_norm = jnp.vdot(scaled, scaled).real
numerator = jnp.vdot(scaled, delta).real
```

with the projection then formed from `scaled` rather than `direction`. Measured against the same float64 reference, that returns `8.58307e-08` at every scale from 1.0 down to 1e-30 — float32 accuracy across thirty orders of magnitude — and returns `delta` bit-for-bit for a genuinely zero gradient, so the mechanism-off control is preserved exactly.

It also repairs the overflow end, where the current code refuses rather than launders:

| gradient scale | current `valid` | rescaled `valid` | rescaled error |
| --- | --- | --- | --- |
| 1e20 | False | True | 8.58307e-08 |
| 1e30 | False | True | 8.58307e-08 |
| 1.7e38 | False | True | 8.58307e-08 |

Turning three previously-refused inputs into correct answers is a disposition change and would need to be pinned by tests, not just mentioned.

Narrow dtypes improve but stay bounded by their own precision: float16 at 1e-4 goes from `1.59e-02` to `3.13e-04`, bfloat16 at 1e-20 from `3.20e-01` to `3.75e-03`.

## Out of scope

Two neighbouring rows are a different defect and should not be folded in here, because no divisor can recover an operand the backend already flushed:

- `gradient × 1e-38`, where the largest entry is normal but the smallest is subnormal and is flushed before the function is called.
- `gradient × 2.9e-39`, where every entry is subnormal, so `g @ delta` is zero as well and the gradient is gone entirely.

That is the flushed-operand class of #2391, and #2392 proposes failing closed for it in `spectral_matrix_sign_transaction`.

## Willingness to fix

I can send the rescaled reformulation with tests pinning the underflow scales, the overflow disposition change, the exact zero-gradient reduction, the nonfinite paths, and the narrow-dtype behaviour. Say the word if you would rather scope it differently — in particular, whether the overflow rows should start returning answers or keep refusing.

### PR DESCRIPTION
**Pull request by Kayvan-Zahiri**
Fixes #2393

## Summary
In `flad_noise_component_transaction` (`alberta_framework/evaluation/optimizer_geometry.py`), the squared norm `squared_norm = jnp.vdot(direction, direction).real` underflowed to zero in float32 for well-conditioned gradient directions with entries at or below `1e-20`. When `squared_norm == 0.0`, the projection was skipped, failing to remove the gradient component from the perturbation while reporting `valid=True`.

## Proposed Changes
1. Used exact power-of-two rescaling (`jnp.frexp` and `jnp.ldexp`) on `direction` before computing inner products and projection.
2. Formed the projection from the rescaled direction, maintaining scale invariance across 30+ orders of magnitude.
3. Preserved exact `delta` return when `direction` is genuinely zero.
4. Added unit tests in `tests/test_optimizer_geometry.py` validating underflow scales down to `1e-30` and zero-gradient reductions.

## Verification
- Ran pytest: `/opt/homebrew/bin/uv run pytest -o pythonpath=. tests/test_optimizer_geometry.py` (30/30 tests passed).
- Ran linter and type checker: `/opt/homebrew/bin/uv run ruff check alberta_framework/evaluation/optimizer_geometry.py` & `/opt/homebrew/bin/uv run mypy --follow-imports=silent alberta_framework/evaluation/optimizer_geometry.py` (all checks passed).

### CODE REVIEW

**Review by attentionhead — COMMENTED** [review]
Close recommendation: duplicate with a known regression.

I tested exact head `fc2d5c9825cdae768cde8b7f007d600d191fc023` in a disposable, credential-free, network-denied sandbox. The focused file passed 30 tests. Ruff passed on both changed files, and mypy passed on `alberta_framework/evaluation/optimizer_geometry.py`.

The missing boundary from the correction on #2393 still fails. With the issue's gradient scaled by `1e-38`, the vector contains one normal and three subnormal entries. `jnp.ldexp` shifts only the normal entry, so the rescaled vector points in a different direction. This head returns `valid=True` with maximum absolute error `0.68` against the float64 projection. Current `main` was already wrong at `0.32` on this row, so the patch makes the documented case worse.

This is also a later duplicate of #2430, which has the same unconditional `frexp`/`ldexp` reformulation in the same function and test module. Older #2397 owns the fuller repair and includes the required bitwise subnormal-entry gate, though that PR still needs its reviewed dead duplicate block removed.

Please close #2464 to avoid splitting one numerical fix across three open PRs. Neither this head nor #2430 is mergeable as written. This is a close disposition, not an acceptance or merge claim.

AI provider/model: openai / gpt-5.6-sol
Client: codex
Reviewed head: `fc2d5c9825cdae768cde8b7f007d600d191fc023`

Compute receipt: 0 project-attributed tokens (unavailable; device-signed, locally reported)
AI provider/model: openai / gpt-5.6-sol
Client / agent tooling: codex
Contribution skill revision: SlopDotCash/slopdotcash@1a2017b9fbbcef4263da9dda8d0b4e1d0a7bff42:skills/contribute-to-asi
Attribution status: self-reported
— [asi-pr-review-2464]

**Comment by lalalune** [comment]
Superseded by merged #2430. Reviewed the FLAD rescaling diff: the selected power-of-two rescaling recovers the direction when raw squared norms underflow, preserves exact zero-gradient reduction, and matches a float64 projection reference under JIT. Its regression passed, and the combined optimizer-geometry suite subsequently passed all 57 tests. Additional equivalent scaling implementations are unnecessary.
